### ISSUE TITLE
Regression in VS 17.7 Preview 1: false positive for CS8602 after call to Assert.IsNotNull (NUnit). No false positive in VS 17.6.

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#68885]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/Regression-in-VS-177-Preview-1:-false-p/10384942)._

---
[severity:I'm unable to use this version]
Hello VS Team!

I need to report a very annoying **regression** in VS 2022 17.7 Preview 1, where false positives for rule `CS8602` (Dereference of a possibly null reference) are reported in the code editor (note: although the code compiles and run successfully), which is not present in VS 2022 17.6 or 17.6.2.

Repro:
- Launch VS 17.6 (note: I'm currently using 17.6.2).
- Create a new NUnit Test Project (for .NET 6). I named mine "RegressionInNullableReferenceTypesAnalysisInUnitTests". For the record, NUnit 3.13.3 is used.
- Set the project option "Treat warnings as errors" on.
- Replace the content of file `UnitTest1.cs` with the following code:
  ```csharp
  #pragma warning disable NUnit2019 // Consider using Assert.That(expr, Is.Not.Null) instead of Assert.IsNotNull(expr)
  #pragma warning disable NUnit2005 // Consider using Assert.That(actual, Is.EqualTo(expected)) instead of Assert.AreEqual(expected, actual)
  
  namespace RegressionInNullableReferenceTypesAnalysisInUnitTests
  {
      public class Tests
      {
          private class Model
          {
              public string? Name { get; set; }
  
              public string? SomeMethod() => Name;
          }
  
          [Test]
          public void Model_SomeMethod()
          {
              var target = new Model { Name = "name" };
  
              string? result = target.SomeMethod();
  
              Assert.IsNotNull(result);
              Assert.AreEqual(4, result.Length); // <-- CS8602 Dereference of a possibly null reference (only in VS 17.7 Preview 1, not in 17.6.2).
          }
      }
  }
  ```
- Note that **no warnings or errors are displayed in the editor**, and the code compiles and the unit test runs successfully.
- Close VS 17.6.*.
- Launch VS 17.7 Preview 1.
- Open the _same_ solution.
- When loading has completed, you now see an **error CS8602 Dereference of a possibly null reference** in the editor on the last `Assert` ==> **regression**!
- Note that the the code still compiles and the unit test still runs successfully ==> **false positive**!

In my actual solution, I have about 5000 unit tests, and opening most of them gives me **tons of false positive errors** in the code editor, which makes **working on unit tests impossible in 17.7 Preview 1**. I'm using 17.6.2 instead.

Note: make sure the solution is not opened by both VS at the same time. Once 17.7 Preview 1 has analyzed the unit test and you close the solution, reopening it in VS 17.6 **might also show the false positive error!** Rebuilding the solution, closing VS 17.6 and reopening it with that solution solves this issue. Probably due to some cache that is shared between both VS.

Please fix this!

---
#### Feedback Bot on 6/7/2023, 00:50 AM: 

(private comment, text removed)
#### Didier Prgomet on 6/7/2023, 01:27 AM: 

(private comment, text removed)
#### Feedback Bot on 6/14/2023, 04:00 PM: 

(private comment, text removed)
#### Didier Prgomet on 7/2/2023, 11:55 PM: 

(private comment, text removed)

---
### Original Solutions
(no solutions)

**Comment by RikkiGibson** [comment]
Assert.IsNotNull in NUnit is not actually attributed such that compiler understands that the method throws if the argument is null. Instead, NUnit uses a diagnostic suppressor to remove nullable warnings cases such as the one in the issue description.

Most likely, there was an issue with this suppressor, possibly due to a bug in Roslyn or due to a bug in the analyzer, which was resolved incidentally at some point.

**Comment by RikkiGibson** [comment]
This might have been fixed by #68270. User noted that the issue was fixed in preview2, when this change was merged.

**Comment by RikkiGibson** [comment]
Thanks for reporting the issue. Feel free to follow up here or a new issue if there are any further questions or concerns.
